ColorPicker: Add an intensity slider to all modes for HDR

beicause opened the pull request:
- Close https://github.com/godotengine/godot-proposals/issues/1031

Close #99366

This PR adopts a relatively simple approach, without storing extra intensity field in Color class.

Adds an intensity slider and limit the max value of RGB to 1 in ColorPicker raw mode. Intensity is inferred from raw RGB and may change after switch color mode. The final color is calculated as `color * pow(2, intensity)`.

Also keep the hex LineEdit always visiable and switch it to code if RGB exceeds the valid range.

Edit: 
<img src=https://github.com/user-attachments/assets/22e94cd7-a22d-41ce-8330-09548698cc1f height=350>

Depends on #102240 to allow greater values.

Edit: 
Now the intensity slider is added to all modes and should work.
I modified the ColorMode and ColorPickerShape so that they only handle colors which the intensity has not been applied.

allenwp commented:
From looking at this, it appears that the intensity slider is the exposure stops (log2/pow2). So an intensity of 0 is +0 stops, which uses the RGB values as-is.

If my understanding is correct, I think this is a little confusing to users who are not familiar with stops. Even though I'm used to working in stops, I was still a bit confused and assumed the intensity slider was simply a multiplier on the RGB values -- This meant that I expected an intensity of 1.0 to result in no modification of the RGB values. (But this is not the case because this is actually +1 stops, not a 1.0 intensity multiplier).

I think having the `+` or `-` indicator on the value would help distinguish that this is stops and not an intensity multiplier. Also, the label should read "stops" or something like that, ideally. Here's an example of how photoshop shows this sort of slider:

![image](https://github.com/user-attachments/assets/79b27ca7-29ba-4e65-ae14-4f7d3eba9843)

beicause commented:
pow 2 is better than multiplier as it's easier to control the appearence of glow. Unity's intensity slider also uses pow 2. If we want to allow negative intensity,  more works need to be done.

allenwp commented:
> If we want to allow negative intensity,  more works need to be done. 

I don't think this is strictly needed, but allowing negative stops is very much a standard convention for this sort of a slider, as shown in the Photoshop example I posted.

beicause commented:
Added support for negative intensity. 
When the HDR threshold of the glow is below 1.0 , adjust the intensity to negative makes sense.

beicause commented:
#102240 adds `get_allow_geater` for sliders. After that, I think the max value of intensity can be 4 for common use.

allenwp commented:
> #102240 adds `get_allow_geater` for sliders. After that, I think the max value of intensity can be 4 for common use.

I was also thinking about whether having the slider max at +6.0 or +4.0. I believe +6.0 is nice because it results in 0.0 being placed in the center of the GUI, which gives a pleasant appearance that reinforces that a value of 0.0 is "neutral" (as in, makes no difference to the colour value). But I agree that in practice, the user shouldn't need to go higher than +4.0...

allenwp commented:
> Added support for negative intensity. When the HDR threshold of the glow is below 1.0 , adjust the intensity to negative makes sense.

This looks great to me, thank you!

Is it possible to use the [prifix](https://docs.godotengine.org/en/stable/classes/class_spinbox.html#class-spinbox-property-prefix) property on the spinbox to add a `+` symbol when `intensity > 0`? Or maybe there is another way to achieve this sort of number formatting?

allenwp reviewed (approved):
My review of the code was not very thorough, but the user experience of this PR looks great to me, now that the intensity slider extends to negative stops and the `+`/`-` signs are shown on the intensity value.

Thanks!

beicause commented:
<img src=https://github.com/user-attachments/assets/f34663e3-8ff2-44af-962b-0098a87c7e2e height=300>

I also colorize the sliders in the RAW mode, like RGB mode. At this point, I think the RAW mode and the RGB mode are similar, maybe we could remove the RAW mode and move the Intensity slider to the RGB mode?

allenwp commented:
> maybe we could remove the RAW mode and move the Intensity slider to the RGB mode?

The power-user part of me likes this, but this would require giving users the option to input colour values in either `255` integer scale or `1.0` float scale in the RGB mode of the colour picker. This might make the RGB mode too cluttered and too much of a visual overload for users who are entirely new to game development and colour picking.

For the reasons of making it accessible to new users, maybe the current approach is fine...

That said, I do have some related thoughts...

My complaint is that RAW mode picks colours in gamma encoding rather than linear, but that's more of an issue with how the `Color` class stores data and there's no way that should be changed... But, maybe there should be an HDR mode that allows colour picking in linear encoding that handles encoding conversions for the user. This is actually the way that Photoshop behaves:

8-bit traditional sRGB colour picker:
![image](https://github.com/user-attachments/assets/4316f5a1-643d-494a-a16c-4a94465fbce3)

32-bit HDR sRGB colour picker:
![image](https://github.com/user-attachments/assets/f8890e8f-821d-4a28-9503-77195ba8b781)

Notice how the HDR colour picker is entirely in linear encoding, which is different than the 8-bit traditional sRGB picker. In this way, maybe the intensity slider should exist in an HDR mode that allows selecting colours in linear encoding rather than being added to a RAW mode (like this PR's current state), since gamma encoded colours don't make sense when working in HDR.

(To be honest, I actually wanted this sort of behaviour before even looking at Photoshop -- I was pleasantly surprised that Photoshop behaved this way, and I think there is value in matching the behaviour of art tools like this...)

So after writing all of this it makes me wonder: maybe the RAW mode should be removed and a new (linear) HDR mode should be added with an intensity slider and automatic linear-to-gamma conversions. The intensity slider would not be added to RGB mode, but an option to input colour values in `1.0` float scale would be added to the RGB mode in a way that wouldn't overload new users with complexity...

Calinou reviewed (commented):
Tested locally, it works as expected.

I find the "stop count" UX more difficult to wrangle around when I want to set precise color values though, like `Color(2, 0, 0)`:

![image](https://github.com/user-attachments/assets/8686d53f-9fbe-481c-acc4-8741998a2c3b)

It seems the exact stop count I need to use is a bit less than `+2.31`, so I can't enter the exact value due to rounding.

This could be alleviated if the raw color field wasn't read-only when it had `Color()` values and you could enter those directly.

allenwp commented:
Thanks for trying out linear/gamma conversions. I like how the colour bars for the sliders show how this is a linear encoding mode.

I believe that the alpha value also must have linear/gamma conversion, but someone please fact check me on this. My use case is setting up an unshaded material as follows:
- Visual shader
- Blend mode: Mix
- Unshaded: true
- ColorConstant Node attached to Albedo and Alpha

When blending, it seems that the blending is performed with gamma values, so an alpha of 0.5 (gamma) blends to colour values of 128/255 (gamma). So if you want control over this blending using linear values, you would need the automatic linear/gamma conversion for alpha, just like the colour channels. But then again... I could be making up a use case that nobody actually wants. I'm uncertain about if my suggestions for linear/gamma conversions will actually be valuable to many people or not.

Additionally, with this behaviour, the "RAW" mode tab must be renamed to "HDR" because it is no longer a RAW editing mode. "RAW" suggests to the user that they are directly editing the raw values for `Color` that are stored in memory and that will be passed to scripted shaders, which is no longer what this colour picker mode does.

I believe that some users would still want the RAW editing mode, especially because script shaders with the `source_color` hint will always provide gamma encoded values. I still don't know what the best way to have a RAW mode is or if there is a clean way to integrate it into the RGB mode without making it feel too cluttered.

I hope that all of this isn't bloating the colour picker too much. It is something that's packed in every default export template, if I'm not mistaken, and so any changes will affect all games/apps that use it...

QbieShay commented:
Hey! I just checked it in engine and I'd like to bring forth 2 requests

1. Can you add the intensity slider to all modes? I personally always use HSV when making VFX, and it'd be annoying to move between raw and HSV all the time
2. I think the up and down arrow should increase the stops by one and not by 0.01

Other than that, everything is great! Thank you for working on this.

beicause commented:
> 1. Can you add the intensity slider to all modes? I personally always use HSV when making VFX, and it'd be annoying to move between raw and HSV all the time
> 2. I think the up and down arrow should increase the stops by one and not by 0.01

I'm OK with these. I will try to acheive it then.

allenwp commented:
I just spoke with an experienced rendering engineer at Lucasfilm and they confirmed that alpha channels, for common purposes such as transparency, are almost always stored and interpreted as linear-encoded values, even when the RGB channels are nonlinear sRGB encoded. That means that the alpha picker for the new HDR linear mode is correct in behaving the same as it does in the RGB nonlinear picker mode.

Because the HDR intensity slider has been added to all modes, the HDR mode's tab could be renamed to "Linear" instead of "HDR", but I'm not sure if that's better or worse than the current behaviour. The intensity slider is what makes a colour picker mode "HDR", so right now, all colour picker modes are HDR modes.

I don't agree that the intensity slider should exist in all colour picker modes because I expect this would be overwhelming to new game developers, but I understand that the user shouldn't need to switch back and forth.

There is still the remaining issue that there is no longer a RAW mode in the colour picker, so this is a regression. One suggestion from Calinou was to allow editing of the script text field. Another idea is to add a "RAW" mode checkbox on the RGB mode or somehing similar. I'm not sure what the best answer... This might need more discussion and mockups to figure out...

beicause commented:
>There is still the remaining issue that there is no longer a RAW mode in the colour picker, so this is a regression.

I followed Calinou's suggestion. The color can be set by executing expression in the script text.

Also, adding the intensity slider to all modes keeps the height of dialog unchanged, which can solve the first issue mentioned in https://github.com/godotengine/godot/pull/99366#pullrequestreview-2446319604

allenwp commented:
> > There is still the remaining issue that there is no longer a RAW mode in the colour picker, so this is a regression.
> 
> I followed Calinou's suggestion. The color can be set by executing expression in the script text, and in this mode the text of label "Hex" will change to "Expr"
> 
> Also, adding the intensity slider to all modes keeps the height of dialog unchanged, which can solve the first issue mentioned in [#99366 (review)](https://github.com/godotengine/godot/pull/99366#pullrequestreview-2446319604)

I'm sorry, my mistake! I did not realize you implemented Calinou's suggestion. Thanks for all of your work on this and responding to my comments; it's good to see the reasoning behind these choices.

My only remaining suggestion is to use `Linear` as the tab label instead of `HDR`, as all colour picker modes now have the HDR intensity slider.

Thanks again, this is looking really great!

KoBeWi commented:
If the extra slider proves to be a problem, it can be made optional via the Customization options.
![image](https://github.com/user-attachments/assets/f2e7bb95-73a2-4410-bdca-7665725c2f5a)

Or as a checkbox in color mode menu:
![image](https://github.com/user-attachments/assets/d116801b-8781-414e-ae43-a6a0d9ec0c63)

beicause commented:
> If the extra slider proves to be a problem, it can be made optional via the Customization options. ![image](https://private-user-images.githubusercontent.com/2223172/425444948-f2e7bb95-73a2-4410-bdca-7665725c2f5a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDI1NjQ3MjMsIm5iZiI6MTc0MjU2NDQyMywicGF0aCI6Ii8yMjIzMTcyLzQyNTQ0NDk0OC1mMmU3YmI5NS03M2EyLTQ0MTAtYmRjYS03NjY1NzI1YzJmNWEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDMyMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAzMjFUMTM0MDIzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MDNhMjRiNzAzYzI5OWYzMGEyNDBmMTU2NWEzYTIyZGVkNzNmNDdmN2JkNTY2ZDAyMTc1YmM4NDdhOGFjNDM5ZSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.EBlu3kZr16Bb9PLoTIFjlSWI8HGqFMNgRiPdwrWeN3k)
> 
> Or as a checkbox in color mode menu: ![image](https://private-user-images.githubusercontent.com/2223172/425445110-d116801b-8781-414e-ae43-a6a0d9ec0c63.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDI1NjQ3MjMsIm5iZiI6MTc0MjU2NDQyMywicGF0aCI6Ii8yMjIzMTcyLzQyNTQ0NTExMC1kMTE2ODAxYi04NzgxLTQxNGUtYWU0My1hNmEwZDllYzBjNjMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDMyMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAzMjFUMTM0MDIzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MGYyZmQwY2YwOTkwZmEzMzZmYTYwYTVmOTdmMjI3NWNhYzEwZmQwZTkwM2I4OTUyYWNjMmY2ZmFjNzk4OTgzNCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.e0Lnbgdjb1kZGeH0-jDUK5LEpdC2RkgSLSGIS3Gfq3k)

I have exposed the `edit_intensity` property, but it cannot be used for the color picker of editor, so I decide to add this option to the color mode menu.

Calinou commented:
> I don't agree that the intensity slider should exist in all colour picker modes because I expect this would be overwhelming to new game developers, but I understand that the user shouldn't need to switch back and forth.

Some colors will ignore HDR values by clipping them (e.g. anything that sets a vertex color in some manner), so the intensity slider should probably not be shown for those either. We'd need new property hints similar to `PROPERTY_HINT_COLOR_NO_ALPHA`.

beicause commented:
>Some colors will ignore HDR values by clipping them (e.g. anything that sets a vertex color in some manner), so the intensity slider should probably not be shown for those either. We'd need new property hints similar to PROPERTY_HINT_COLOR_NO_ALPHA.

Are there any properties that clip HDR color? It seems there are no properties for setting vertex colors.

Calinou commented:
> Are there any properties that clip HDR color? I'm not sure if there are any properties for vertex color that clip HDR color.

Vertex colors in Godot 4 are stored as RGBA8, so they can't contain HDR data. This affects properties like Sprite3D Modulate (although https://github.com/godotengine/godot/pull/97604 can work around it).

KoBeWi commented:
> I have exposed the edit_intensity property, but it cannot be used for the color picker of editor, so I decide to add this option to the color mode menu.

It's a bit redundant to have both menu option and a property. The menu option could be removed and you can make editor setting instead. There is `setup_color_picker()` method that setups all ColorPickers in the editor.

beicause commented:
> Also Hex display is wrong with intensity:
> 
>  godot.windows.editor.dev.x86_64_gKcnbFD0lj.mp4 
> (though it's still better than on master, where the LineEdit just disappears...)

It's intentional, overbright colors can't be represented as hex.

allenwp commented:
> > Also Hex display is wrong with intensity:
> > godot.windows.editor.dev.x86_64_gKcnbFD0lj.mp4
> > (though it's still better than on master, where the LineEdit just disappears...)
> 
> It's intentional, overbright colors can't be represented as hex.

At the end of the video there seems to be an issue with the HEX/Expr button still being active when only Expr is valid; it can display the “HEX” label when an expression is shown.

beicause commented:
The "Hex" label indicates that the LineEdit will revert to hex mode when the color is no longer overbright. The button still works.

Alternatively, we can disable it when the color is overbright.

allenwp commented:
> Alternatively, we can disable it when the color is overbright.

I think from a usability perspective it makes sense to disable the button when the primary effect of clicking the button (viewing of hex representation) cannot be seen.

allenwp commented:
Just tried the latest (pre-squash) artifact and the math and behaviour is as expected for the most common usage. I haven't looked into the edge cases that others have brought up previously.

The new RAW input behaviour (typing into the expression field) may be unexpected to those who are familiar with the existing RAW mode, so it may be worthwhile to highlight this change in the release notes. I am not sure where in the online documentation this behaviour should be explained.

Additionally, the new Linear colour picker mode (as well as the OKHSL mode) should be added to [this part of the documentation](https://docs.godotengine.org/en/stable/about/list_of_features.html#gui-system) after this PR has been merged:

Existing text:
> Color picker with RGB and HSV modes.

Proposed text:
> Color picker with RGB, HSV, Linear, and OKHSL modes.

KoBeWi commented on editor/editor_settings.cpp:
```diff
@@ -558,6 +558,7 @@ void EditorSettings::_load_defaults(Ref<ConfigFile> p_extra_config) {
 
 	EDITOR_SETTING_BASIC(Variant::INT, PROPERTY_HINT_ENUM, "interface/inspector/default_color_picker_mode", (int32_t)ColorPicker::MODE_RGB, "RGB,HSV,RAW,OKHSL")
 	EDITOR_SETTING_BASIC(Variant::INT, PROPERTY_HINT_ENUM, "interface/inspector/default_color_picker_shape", (int32_t)ColorPicker::SHAPE_OKHSL_CIRCLE, "HSV Rectangle,HSV Rectangle Wheel,VHS Circle,OKHSL Circle")
+	EDITOR_SETTING_BASIC(Variant::BOOL, PROPERTY_HINT_NONE, "interface/inspector/default_color_picker_show_intensity", true, "");
```
`default` does not really fit here.
The 2 properties' "default" refers to mode and shape.

KoBeWi commented on scene/gui/color_picker_shape.cpp:
```diff
@@ -122,7 +122,7 @@ void ColorPickerShape::draw_sv_square(Control *p_control, const Rect2 &p_square,
 	};
 
-	Color color1 = color_picker->color;
+	Color color1 = color_picker->_get_pick_color_normalized();
```
```suggestion
	Color color1 = color_picker->color_normalized;
```
Shapes have access to picker's private members.

beicause commented on scene/gui/color_picker.cpp:
```diff
@@ -879,7 +1009,7 @@ inline int ColorPicker::_get_preset_size() {
 	ColorPresetButton *btn_preset_new = memnew(ColorPresetButton(p_color, p_size));
 	btn_preset_new->set_tooltip_text(vformat(atr(ETR("Color: #%s\nLMB: Apply color\nRMB: Remove preset")), p_color.to_html(p_color.a < 1)));
-	btn_preset_new->set_accessibility_name(vformat(atr(ETR("Color: #%")), p_color.to_html(p_color.a < 1)));
+	btn_preset_new->set_accessibility_name(vformat(atr(ETR("Color: #%s")), p_color.to_html(p_color.a < 1)));
```
"%s" is missing here.
In addition, perhaps we should use the values of Color instead of `to_html()`. This will always return "ff" for overbright components.

beicause commented:
I found that after enabling HDR 2D, the gradient drawn by draw_polygon will  be linear interpolation. So I have to also use GradientTexture2D in rgb mode. (hsv and okhsl are also affected, but less obvious)

KoBeWi commented on scene/gui/color_mode.cpp:
```diff
@@ -66,37 +66,47 @@ void ColorModeRGB::slider_draw(int p_which) {
+	Ref<Gradient> gradient = gradient_texture->get_gradient();
+	if (gradient.is_null()) {
+		gradient.instantiate();
+		PackedFloat32Array offsets;
+		offsets.resize(2);
+		offsets.write[0] = 0;
+		offsets.write[1] = 1;
```
```suggestion
		PackedFloat32Array offsets = {0, 1};
```

KoBeWi commented on scene/gui/color_mode.cpp:
```diff
@@ -66,37 +66,47 @@ void ColorModeRGB::slider_draw(int p_which) {
-	pos.set(1, Vector2(size.x, 0));
-	pos.set(2, Vector2(size.x, margin));
-	pos.set(3, Vector2(0, margin));
+	PackedColorArray colors;
+	colors.resize(2);
+	colors.write[0] = left_color;
+	colors.write[1] = right_color;
 
```
```suggestion
	PackedColorArray colors = {left_color, right_color};
```

KoBeWi commented on scene/gui/color_picker.cpp:
```diff
@@ -452,34 +512,78 @@ void ColorPicker::_slider_value_changed() {
+		v = col_normalized.get_v();
+	}
+	color_normalized = col_normalized;
+	float alpha = alpha_value->get_value() / (current_mode == MODE_LINEAR ? 1 : 255);
```
Why alpha isn't returned by color mode?
Ideally we should avoid checking for specific color mode in ColorPicker, because abstracting it out is their purpose. I only just noticed that we still have such checks, this code needs refactoring.

beicause replied:
Intensity is the fourth slider, while alpha is the fifth. It's a bit strange to put alpha into the color_mode and process it as the fourth one.

KoBeWi replied:
Alpha depends on color mode though (as seen in the line above), so it needs to be handled by color mode.
Also intensity is the strange one if anything. Alpha is part of the color, intensity isn't.

beicause commented:
I have integrated #105729 into this PR in advance because I have sufficient bandwidth to do this now.

KoBeWi commented on scene/gui/color_mode.cpp:
```diff
@@ -66,37 +85,40 @@ void ColorModeRGB::slider_draw(int p_which) {
+		rgb_texture[p_which]->set_width(400);
+		rgb_texture[p_which]->set_height(6);
+	}
+	gradient_texture = rgb_texture[p_which];
```
```suggestion
	Ref<GradientTexture2D> gradient_texture = rgb_texture[p_which];
```
The above definition is unnecessary.

KoBeWi commented on scene/gui/color_mode.cpp:
```diff
@@ -108,47 +130,51 @@ void ColorModeHSV::_value_changed() {
+	color_picker->s = color_picker->sliders[1]->get_value() / 100.0;
+	color_picker->v = color_picker->sliders[2]->get_value() / 100.0;
+
+	color_picker->hsv_cache_once = true;
```
Shouldn't this be `okhsl_cache_once`? You are setting the hsv above.

beicause replied:
The naming may be misleading, but it should set hsv_cache_once here. See https://github.com/godotengine/godot/blob/fce71de41914781a87e5a786a78ca04a22700ae4/scene/gui/color_picker.cpp#L648-L661

KoBeWi commented on scene/gui/color_picker.cpp:
```diff
@@ -360,6 +384,16 @@ void ColorPicker::_update_controls() {
 		alpha_label->hide();
 	}
 
+	if (edit_intensity) {
+		intensity_value->show();
+		intensity_slider->show();
+		intensity_label->show();
```
This can be changed to
```suggestion
	intensity_value->set_visible(edit_intensity);
	intensity_slider->set_visible(edit_intensity);
	intensity_label->set_visible(edit_intensity);
```

KoBeWi commented on scene/gui/color_picker.cpp:
```diff
@@ -401,7 +435,7 @@ void ColorPicker::_set_pick_color(const Color &p_color, bool p_update_sliders) {
 
 void ColorPicker::set_pick_color(const Color &p_color) {
-	_set_pick_color(p_color, true); //because setters can't have more arguments
+	_set_pick_color(p_color, true, true); //because setters can't have more arguments
```
```suggestion
	_set_pick_color(p_color, true, true); // Because setters can't have more arguments.
```

beicause committed:
ColorPicker: Add intensity slider

Add intensity slider to all color modes. Replace raw mode by linear mode, which uses linear color space.

When color is overbright, automatically switch hex text to script text. Allow executing expression in script text field to set color. Add the "script" icon to the default theme.
